Network packet socket support. See "man 7 packet" on Linux.

rveerama1 opened the pull request:
Initial version of packet socket support to Networking subsystem. This PR is targeted for basic feature, like recv and send every packet on the wire (i.e AF_PACKET, SOCK_RAW and ETH_P_ALL). It means user's of socket API has to include all the headers (L2, L3 and so on) while sending, also the will receive all the packets including L2 header. This version supports raw socket over Ethernet only. Also combination of other family and protocols not supported in this version.

jukkar committed:
net: ethernet: Add more IEEE 802.3 protocol types

Add ETH_P_xxx protocol types if they are missing. After this
we can use the protocol types when working with BSD sockets.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: ip: Add helpers for packet socket support

Various defines and helpers for supporting packet sockets.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>
Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
net: pkt: Set default data_len to iface MTU

When creating net_pkt, the default value of net_pkt data_len was
set to zero. In case of RAW sockets, when family is AF_PACKET,
the data_len will be zero as there is no higher level protocol
information. So in this case, the default value of data_len
must be set to interface MTU

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
net: sockets: Add socket api to support AF_PACKET

This commit adds packet socket support to socket api.
This version supports basic packet socket features.
Protocol family is AF_PACKET, type of socket is
SOCK_RAW and proto type is ETH_P_ALL. The user will
receive every packet (with L2 header) on the wire.
For TX, the subsystem expects that the user has set
all the protocol headers (L2 and L3) properly.

Networking subsystem doesn't verify or alter the headers while
sending or receiving the packets. This version supports packet
socket over Etherent only. Also combination of other family
and protocols doesn't work (i.e. Application can not open
packet-socket and non packet-socket together).

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

tbursztyka commented on subsys/net/ip/net_core.c:
```diff
@@ -65,12 +66,66 @@ void net_analyze_stack(const char *name, const char *stack, size_t size)
+#if defined(CONFIG_NET_SOCKETS_RAW)
+static inline void process_raw_data(struct net_pkt *pkt)
+{
+	struct net_context *cont;
```
s/cont/ctx

tbursztyka commented on subsys/net/ip/net_core.c:
```diff
@@ -65,12 +66,66 @@ void net_analyze_stack(const char *name, const char *stack, size_t size)
+		return;
+	}
+
+	raw_pkt = net_pkt_clone(pkt, K_MSEC(100));
```
Use a macro for the timeout, as everywhere in the core stack. Something like BUF_TIMEOUT

tbursztyka commented on subsys/net/ip/net_core.c:
```diff
@@ -65,12 +66,66 @@ void net_analyze_stack(const char *name, const char *stack, size_t size)
+		/* TODO Loolback packets */
+		process_raw_data(pkt);
+
+		return NET_DROP;
```
So as soon as CONFIG_NET_SOCKETS_RAW is enabled, all packet will be treated as raw stuff, and dropped?

tbursztyka commented on subsys/net/ip/net_stats.h:
```diff
@@ -287,6 +287,152 @@ static inline void net_stats_update_per_proto_drop(struct net_if *iface,
+	}
+}
+
+#if defined(CONFIG_NET_STATISTICS_RPL)
```
Update your branch: RPL is no more.

pfalcon reviewed (changes requested):
Please explain why https://github.com/zephyrproject-rtos/zephyr/pull/11229 was closed and this submitted instead? What improvements were made comparing to #11229? There was enough comments on #11229, please go over them and respond which were addresses and which weren't.

tbursztyka commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -500,7 +500,16 @@ static int ethernet_send(struct net_if *iface, struct net_pkt *pkt)
 	int ret;
 
-	if (IS_ENABLED(CONFIG_NET_IPV4) &&
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_RAW) &&
```
Insert this right after ipv6

rveerama1 commented on subsys/net/ip/net_core.c:
```diff
@@ -65,12 +66,66 @@ void net_analyze_stack(const char *name, const char *stack, size_t size)
+		return;
+	}
+
+	net_pkt_set_family(raw_pkt, AF_PACKET);
```
@tbursztyka Important issue is here. When core receives packet from driver, core will call L2 to verify header and remove L2 header. In case of raw socket, we have to just call L2 to verify and not to exclude.
Second issue is, If user opens a raw socket and non-raw socket. Then same packet has to be feed to raw socket with L2 header and for non-raw socket only data has to be feed.
Do we need to clone or something else? any suggestions?

tbursztyka commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -500,7 +500,16 @@ static int ethernet_send(struct net_if *iface, struct net_pkt *pkt)
+			net_pkt_unref(pkt);
+		}
+
+		return ret;
```
You are not updating eth stats here, 

so don't send here, just a goto that goes to line 571 and that's it

pfalcon reviewed (changes requested):
-1 for patching any existing samples. Samples are samples - easy pieces of code for people to quickly learn from, not scare them away with walls of code.

Instead, please act on the comment made in #11229 - provide sample(s) which show that the same code works both in Linux and in Zephyr.

rveerama1 commented:
> Please explain why #11229 was closed and this submitted instead? What improvements were made comparing to #11229? There was enough comments on #11229, please go over them and respond which were addresses and which weren't.

@pfalcon Jukka submitted this PR 2 months ago and he was on vacation on two different time intervals. I am trying to address comments from old PR too.
Removal of connection_raw.c and separate patch for net_stats.c
A sample which works on linux and zephyr.
Instead of feeding code in every driver for raw socket related things, something common.

rveerama1 commented:
> -1 for patching any existing samples. Samples are samples - easy pieces of code for people to quickly learn from, not scare them away with walls of code.

ATM really not paying attention to -1 or +1. Look at my label DNM. I am not in hurry to merge this. I am just concentrating on architecture and design changes. How it would fit with current ongoing major changes from @tbursztyka

pfalcon commented:
> What improvements were made comparing to #11229?

Well, I can see at least the following: this actually implements RAW *sockets* (i.e. exposes via BSD Sockets API), and does so uses recently introduced struct socket_op_vtable. That's definitely the right way, thanks for using it. And the integration looks pretty clean, as expected. (Will need to look in more detail yet.)

pfalcon commented:
@rveerama1 
> I am trying to address comments from old PR too. ...

Good, thanks.

> ATM really not paying attention to -1 or +1. ... I am just concentrating on architecture and design changes.

Sounds good, just trying to note high-level things which may need improvement.

rveerama1 committed:
samples: sockets: Add packet socket sample application

Sample application which opens a packet socket and receives
every packet on the wire and send some dummy packet over
socket. Simple demo of how to use packet sockets.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 commented:
Dropped unnecessary patches from original PR. So far no use at the moment. Just in case fyi dropped patches are (drivers: eth: native_posix: Add raw socket support , net: pkt: Add helpers to set/get the L2 protocol).

rveerama1 commented:
Dropped another patch [net: ethernet: Add more IEEE 802.3 protocol types].

rveerama1 committed:
net: ip: Add protocol family type to conn handler

As we are adding more protocol families and protocol types
to connection handlers, some values might be same across
different types. Current connection handler only stores
proto type to match the handler, which is not enough if
we add more types. Also combination of family and types
may vary too. So adding family to connection handler to
figure out best match.

Also changing proto variable in net_conn from u8_t to u16_t.
net_context has 16 bit proto.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
net: context: Add packet socket support

This commit adds basic packet socket support to net_context and
allows application to receive or send network packets in raw
format.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>
Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

jukkar commented:
ping @pfalcon can you take a look again to this, the PR is needed for socket CAN support.

pfalcon commented:
> ping @pfalcon can you take a look again to this, the PR is needed for socket CAN support.

Well, I just added similar comments to CAN PR, but first of all it applies here of course: can we get some normative references to what exact API is being implemented (linking to POSIX, Linux docs, whatever) into PR description and/or commit messages?

I'll look in detail later today.

(I'm still not sure that we should haste with this for 1.14.)

rveerama1 commented:
No major functional changes, fixing missing bits and pieces. (documentation, commit messages, cleaning patches from unrelated changes and etc.

pfalcon commented on include/net/net_context.h:
```diff
@@ -283,11 +294,14 @@ struct net_context {
 
-	/** Network interface assigned to this context */
-	u8_t iface;
+	/** Protocol (UDP, TCP or IEEE 802.3 protocol value) */
```
What is "IEEE 802.3 protocol value"? What if value of "IEEE 802.3 protocol value" clashes with value for UDP or TCP?

pfalcon commented on include/net/net_context.h:
```diff
@@ -349,41 +363,52 @@ static inline void net_context_set_state(struct net_context *context,
  * @brief Set address family for this network context.
  *
- * @details This function sets the address family (IPv4 or IPv6)
+ * @details This function sets the address family (IPv4, IPv6, packet)
```
> This function sets the address family (IPv4, IPv6, packet)

"packet address family" sounds strange to me. Suggestion: write "(IPv4, IPV6, etc.)" or remove that part altogether.

jukkar replied:
What is strange with AF_PACKET? No need to change anything here.

pfalcon replied:
Which AF_PACKET? There's no "AF_PACKET", there's "packet", and it's strange, yeah.

pfalcon commented on include/net/net_context.h:
```diff
@@ -349,41 +363,52 @@ static inline void net_context_set_state(struct net_context *context,
-		context->flags |= NET_CONTEXT_FAMILY;
-		return;
+	if (family == AF_INET) {
+		flag = NET_CONTEXT_AF_INET << 4;
```
Is this "4" magic constant really belongs here?

pfalcon commented on include/net/net_context.h:
```diff
@@ -414,36 +445,32 @@ enum net_sock_type net_context_get_type(struct net_context *context)
+	u16_t flag = 0;
+
+	if (type == SOCK_DGRAM) {
+		flag = NET_CONTEXT_SOCK_DGRAM << 6;
```
Another magic constant. Why 6, why not 5?

pfalcon commented on include/net/net_ip.h:
```diff
@@ -40,18 +40,24 @@ extern "C" {
 	IPPROTO_UDP = 17,
 	IPPROTO_ICMPV6 = 58,
+	IPPROTO_RAW = 255, /* As per IANA 255 is Reserved,
+			    * but RFC 3542 recommends 255.
```
Ok, here we go, first normative reference! RFC 3542 is "Advanced Sockets Application Program Interface (API) for IPv6" and it indeed includes a chapter on raw sockets: https://tools.ietf.org/html/rfc3542#section-3

This information should go into PR description, to help reviewers, and in commit message(s).

pfalcon commented on include/net/net_ip.h:
```diff
@@ -128,14 +135,65 @@ struct sockaddr_in_ptr {
+};
+
+/* Packet types.  */
+#define PACKET_HOST         0     /* To us            */
```
Is it really "to us", or "to single host"?

pfalcon commented on samples/net/sockets/raw/Kconfig:
```diff
@@ -0,0 +1,49 @@
+# Kconfig - Private config options for echo-server sample app
```
Comment should be fixed?

pfalcon commented on samples/net/sockets/raw/README.rst:
```diff
@@ -0,0 +1,27 @@
+This sample is a simple RAW socket application showing usage of
+raw socket over ethernet. Sample prints every packet which
+etherent receives and sends every packet over the wire. Zephyr
+network subsystem does not touch any of the headers(L2, L3 and
```
Space after "headers"

pfalcon commented on samples/net/sockets/raw/README.rst:
```diff
@@ -0,0 +1,27 @@
+********************
+
+When the application runs, it opens a raw socket and prints
+the length of packet it receives. And sends dummy packet over
```
So. it sends not "every packet" after all, but a dummy one every 5s. Re: https://github.com/zephyrproject-rtos/zephyr/pull/12564/files#r251889978

pfalcon commented on samples/net/sockets/raw/README.rst:
```diff
@@ -0,0 +1,27 @@
+See the `net-tools`_ project for more details.
+
+This sample can be built and executed on QEMU as described
+in :ref:`networking_with_qemu`.
```
I guess it's not enough to tell just that. I guess it should say that outgoing packets can be observed with Wireshark (or if there's "easier" way, it should mention that).

pfalcon commented on samples/net/sockets/raw/prj.conf:
```diff
@@ -0,0 +1,44 @@
+CONFIG_NET_SHELL=y
+
+# Raw socket configuration
+CONFIG_NET_SOCKETS_RAW=y
```
I'd say this should go to "Generic networking options". Otherwise, there you disable IPv4/IPv6, but don't "enable" anything, looks strange.

pfalcon commented on samples/net/sockets/raw/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+tests:
+  test:
+    harness: net
+    platform_whitelist: native_posix
```
That's too little I guess.

pfalcon commented on subsys/net/ip/connection.c:
```diff
@@ -457,6 +458,14 @@ static int find_conn_handler(enum net_ip_protocol proto,
+			continue;
+		}
+
+		if (conns[i].family != family) {
```
Duplicated chunk of code.

pfalcon commented on subsys/net/ip/connection.c:
```diff
@@ -844,8 +854,15 @@ enum net_verdict net_conn_input(enum net_ip_protocol proto, struct net_pkt *pkt)
+		src_port = dst_port = 0;
 	} else {
-		NET_DBG("No UDP or TCP configured, dropping packet.");
+		NET_DBG("No UDP or TCP or SOCK_RAW configured,"
```
Please replace with "No suitable protocol handler configured ..."

pfalcon commented on subsys/net/ip/net_context.c:
```diff
@@ -619,6 +708,12 @@ int net_context_listen(struct net_context *context, int backlog)
 
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_RAW) &&
+	    net_context_get_family(context) == AF_PACKET) {
+		net_context_set_state(context, NET_CONTEXT_LISTENING);
```
Can RAW sockets be listening at all? UDP sockets can be listening, and RAW is even more hardcore than that.

pfalcon commented on subsys/net/ip/net_context.c:
```diff
@@ -876,7 +976,8 @@ int net_context_accept(struct net_context *context,
 
 	if ((net_context_get_state(context) != NET_CONTEXT_LISTENING) &&
-	    (net_context_get_type(context) != SOCK_STREAM)) {
+	    (net_context_get_type(context) != SOCK_STREAM &&
```
Similar question - can RAW sockets accept() at all?

rveerama1 replied:
No. There is no client and server role in raw socket to my understanding.

pfalcon replied:
So again, this marked as resolved, but isn't resolved. There should be just no this change in the PR.

rveerama1 replied:
sorry, I missed it.

pfalcon commented on subsys/net/ip/net_context.c:
```diff
@@ -1239,6 +1388,67 @@ static int recv_udp(struct net_context *context,
+	/* If there is no callback registered, then we can only drop
+	 * the packet.
+	 */
+
```
Stray empty line.

pfalcon replied:
Still an issue.

rveerama1 replied:
ok

pfalcon commented on subsys/net/ip/net_context.c:
```diff
@@ -1378,6 +1594,36 @@ int net_context_get_option(struct net_context *context,
 	return ret;
 }
 
+struct net_context *net_context_find(sa_family_t family,
```
Such a function doesn't exist already? How have we survived for so long without it?

rveerama1 replied:
I think we have some confusion here. This function is added from the beginning in my PR. Only thing I split the patch. Tomasz was asking to split some patches because of various changes.

pfalcon replied:
> This function is added from the beginning in my PR. Only thing I split the patch.

Sounds good. But can you summarize in the commit message why we need such a function?

rveerama1 replied:
https://github.com/zephyrproject-rtos/zephyr/pull/12564/commits/77880b96127455a6619e1bf172182c504bb4881d This helper need by this patch.

pfalcon replied:
> 77880b9 This helper need by this patch.

So, please use a commit message like:

~~~
Helper finds the matching network context based on family,
type and protocol: net_context_find(). This function will be
used in the implementation of AF_PACKET net_context, which
are by their nature "connectionless", so it's not convenient to use
existing struct net_conn machinery.
~~~

Btw, I still don't understand why you don't reuse net_conn, especially **after** you patched it to accept address family, but I don't go as far as thoroughly questioning that matter in the current situation (the PR being rushed for 1.14).

rveerama1 replied:
Hmm, after all this API is not required at the moment. It was introduced for a reason. If user open a packet socket and udp/tcp socket. The same net_pkt has to be feed to both the sockets. So if user enabled CONFIG_NET_SOCKETS_PACKET and opened a packet socket, then packet has to be feed as it it. Then for udp socket data has to feed. May be we have to clone the packet. But If user enabled CONFIG_NET_SOCKETS_PACKET but doesn't opened any packet socket, then we need not to clone. So to find that particular context we need this helper.

After all in this initial version we don't need this at all. I will remove that patch.

pfalcon commented on subsys/net/ip/net_core.c:
```diff
@@ -29,6 +29,7 @@ LOG_MODULE_REGISTER(net_core, CONFIG_NET_CORE_LOG_LEVEL);
 #include <net/tcp.h>
 #include <net/gptp.h>
 #include <net/lldp.h>
+#include <net/ethernet.h>
```
Ethernet in net_core? Oh, no! @tbursztyka did so good work to separate L2's from generic networking and it creeps back now?

jukkar replied:
Only to get ETH_P_ALL which is used in raw sockets to inform that the caller wants to receive all packets.

pfalcon replied:
> Only to get ETH_P_ALL which is used in raw sockets to inform that the caller wants to receive all packets.

Not sure if it's needed. That header can be included in sockets_packet.c, but I don't see a good reason why it should be included in net/ip/net_core.c.

rveerama1 replied:
Provided separate file raw.c and raw.h

pfalcon commented on subsys/net/ip/net_core.c:
```diff
@@ -65,12 +66,47 @@ void net_analyze_stack(const char *name, const char *stack, size_t size)
+	/* Currently we are skipping L2 layer verification and not
+	 * removing L2 header from packet.
+	 * TODO :
+	 * 1) Pass it through L2 layer, so that L2 will verifies
```
"will verifies" has a typo

pfalcon commented on subsys/net/ip/net_core.c:
```diff
@@ -65,12 +66,47 @@ void net_analyze_stack(const char *name, const char *stack, size_t size)
+	 * TODO :
+	 * 1) Pass it through L2 layer, so that L2 will verifies
+	 * that packet is intended to us or not and sets src and dst lladdr.
+	 * And L2 should not pullof L2 header when combination of socket
```
"pullof" is a typo

rveerama1 replied:
fixed

pfalcon commented on subsys/net/ip/net_core.c:
```diff
@@ -65,12 +66,47 @@ void net_analyze_stack(const char *name, const char *stack, size_t size)
+		return NET_DROP;
+	}
+
+	/* Currently we are skipping L2 layer verification and not
```
This comment is golden, and should go into the PR description and into commit message.

rveerama1 replied:
Tried to update as many places I can in new version.

pfalcon commented on subsys/net/ip/net_stats.h:
```diff
@@ -287,6 +287,52 @@ static inline void net_stats_update_per_proto_drop(struct net_if *iface,
 	}
 }
 
+static inline void net_stats_update_l2_proto_recv(struct net_if *iface,
```
@tbursztyka, what would say about this? Don't we have too many stats layers. Like, L3 stats, ethernet stats, and now these L2 stats.

tbursztyka replied:
I'd leave the stats out of the picture for now. (in a separate and future patch, if needed)

rveerama1 replied:
Stats dropped in new version. I will work on it later.

pfalcon commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -96,6 +96,12 @@ config NET_SOCKETS_OFFLOAD
+	bool "Enable raw socket support"
+	default n
+	help
+	  The value depends on your network needs.
```
So, here will go description of what flavors of RAW sockets we actually support.

rveerama1 replied:
Added more text in new version.

pfalcon commented on subsys/net/lib/sockets/sockets_raw.c:
```diff
@@ -0,0 +1,351 @@
+	return zraw_bind_ctx(obj, addr, addrlen);
+}
+
+/* The connect() function is no longer necessary. */
```
What do you mean "no longer necessary"? It either should work the same way as for SOCK_DGRAM, storing a "default" address to let send() work, or return an error. How does your Linux behave? (Assuming you can't look up a normative reference for that.)

rveerama1 replied:
As per my understanding in this version its not needed. Let's add in the future if there any requirement for it.

pfalcon commented on subsys/net/lib/sockets/sockets_raw.c:
```diff
@@ -0,0 +1,351 @@
+}
+
+/*
+ * The listen() and accept() functions are without any functionality,
```
If they're "without any functionality", why did you implement them on net_context level?

pfalcon replied:
Wasn't addressed.

pfalcon commented:
Commit message:

>  net: pkt: Set default data_len to iface MTU
> Set default value if data_len to interface MTU size than 0.

2nd sentence is not legible.

pfalcon commented:
> can we get some normative references to what exact API is being implemented (linking to POSIX, Linux docs, whatever) into PR description and/or commit messages?

Ok, I tried to investigate this matter myself from the current state of the PR:

1. Commit titled "net: sockets: Add socket api to support SOCK_RAW" has more or less comprehensive description of what subset of "raw socket" functionality this PR tries to implement. Well, actually the biggest point is exactly that this implements **subset** of raw socket functionality. However, that commit is too informal IMHO. No references to normative docs defining raw sockets in general, and this particular subset. No description of motivation to implement exactly this subset, instead of more generic raw socket support.
2. Code comment in some place refers to https://tools.ietf.org/html/rfc3542#section-3 , which is already something, thanks. Should be made more prominent.
3. https://github.com/zephyrproject-rtos/zephyr/pull/12564/files/0a66c5cbbed8fef5f9c7dbee55a4af3b7d19617f#diff-3a9f6b44fa4a4f8b7b164c11d83b183aR79 provides some additional info what subset is implemented, and what's not implemented. Again, no references.

So, that all at least something, but I would like to start from start. The POSIX standard is what defines sockets API, so let's see what it has to say about raw sockets.

Turns out, not that much:

1. http://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_10_06 says:

> [RS] [Option Start] The SOCK_RAW socket type is similar to the SOCK_DGRAM type. It differs in that it is normally used with communication providers that underlie those used for the other socket types. For this reason, the creation of a socket with type SOCK_RAW shall require appropriate privileges. The format of datagrams sent and received with this socket type generally include specific protocol headers, and the formats are protocol-specific and implementation-defined. [Option End]

As can be seen, SOCK_RAW description is guarded by "RS" option, which is described (view-source:http://pubs.opengroup.org/onlinepubs/9699919799/help/codes.html#RS) as:

> [RS] [Option Start] Raw Sockets [Option End]
The functionality described is optional. The functionality described is also an extension to the ISO C standard.

> Where applicable, functions are marked with the RS margin legend in the SYNOPSIS section. Where additional semantics apply to a function, the material is identified by use of the RS margin legend.

What's more: http://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_10_18 says:

> When a socket is created in the Internet family with a protocol value of zero, the implementation shall use the protocol listed below for the type of socket created.
> SOCK_RAW
    [RS] [Option Start] IPPROTO_RAW. [Option End]

> [RS] [Option Start] A raw interface to IP is available by creating an Internet socket of type SOCK_RAW. The default protocol for type SOCK_RAW shall be identified in the IP header with the value IPPROTO_RAW. Applications should not use the default protocol when creating a socket with type SOCK_RAW, but should identify a specific protocol by value. The ICMP control protocol is accessible from a raw socket by specifying a value of IPPROTO_ICMP for protocol. [Option End]

That's literally all hits for "SOCK_RAW" in POSIX: http://pubs.opengroup.org/cgi/susv4search.cgi?KEYWORDS=SOCK_RAW&SUBSTRING=substring&CONTEXT= 

In other words, POSIX specifies that SOCK_RAW may be (optionally) part of sockets API, but doesn't really specify any semantics.

POSIX doesn't have any hits for "AF_PACKET".

pfalcon commented:
So, what's that AF_PACKET then? Let's see http://man7.org/linux/man-pages/man7/packet.7.html :

> packet_socket = socket(AF_PACKET, int socket_type, int protocol);

Whoa, no explicit mentioning of SOCK_RAW here!

> Packet sockets are used to receive or send raw packets at the device
       driver (OSI Layer 2) level.

Aha, "packet sockets", not raw sockets.

> The socket_type is either SOCK_RAW for raw packets including the
       link-level header or SOCK_DGRAM for cooked packets with the link-
       level header removed.

That clarifies it.


So, I have a strong suspicion that this PR instead of "raw sockets" implements "packet sockets". If so, please rename the PR and all references in it accordingly.

pfalcon commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -119,6 +119,12 @@ int _impl_zsock_socket(int family, int type, int proto)
 #endif
 
+#if defined(CONFIG_NET_SOCKETS_RAW)
+	if (type == SOCK_RAW) {
```
Per the revelation in https://github.com/zephyrproject-rtos/zephyr/pull/12564#issuecomment-458646644 (challenges to it are welcome), this should dispatch on `family == AF_PACKET` instead.

rveerama1 replied:
Done.

tbursztyka commented on subsys/net/ip/net_core.c:
```diff
@@ -65,12 +66,47 @@ void net_analyze_stack(const char *name, const char *stack, size_t size)
+	return net_conn_input(ETH_P_ALL, pkt);
+}
+#endif
+
```
just create a raw.c (as it exists an ipv4.c, ipv6.c ) and put that function there (named net_raw_input() )
Have a raw.h, which one will have ifdefs around CONFIG_NET_SOCKETS_RAW, this will at least solve the inclusion ethernet.h out of net_core.c 
Also, we might end up with more code for raw stuff in the future so raw.c will be the place.

rveerama1 replied:
Done.

tbursztyka commented on subsys/net/ip/net_core.c:
```diff
@@ -65,12 +66,47 @@ void net_analyze_stack(const char *name, const char *stack, size_t size)
 	bool locally_routed = false;
 
+#if defined(CONFIG_NET_SOCKETS_RAW)
+	return process_raw_data(pkt);
```
I still don't get this, so we can't have SOCK_STREAM, SOCK_DGRAM (and thus ipv4, ipv6, udp, tcp....) at same time as SOCK_RAW?

rveerama1 replied:
Yes, We can not open raw and non-raw socket at the moment. I will be working on it later.

tbursztyka commented on include/net/net_context.h:
```diff
@@ -45,21 +45,30 @@ enum net_context_state {
 	NET_CONTEXT_LISTENING = 3,
 };
 
+enum {
```
Too many things in that patch. I'd slice in incremental patches: first context, then core, then ethernet l2.

rveerama1 replied:
I split as many as I can in new version.

tbursztyka commented on subsys/net/lib/sockets/CMakeLists.txt:
```diff
@@ -7,5 +7,6 @@ zephyr_sources(
   sockets_select.c
   )
 zephyr_sources_ifdef(CONFIG_NET_SOCKETS_SOCKOPT_TLS sockets_tls.c)
+zephyr_sources_ifdef(CONFIG_NET_SOCKETS_RAW sockets_raw.c)
```
there is an untold "rule": source file are indented to the same column generally (as socket_offload.c below)

rveerama1 replied:
No idea why its like that.

pfalcon commented:
> So, I have a strong suspicion that this PR instead of "raw sockets" implements "packet sockets".

I would consider this confirmed:

* http://man7.org/linux/man-pages/man7/raw.7.html "Raw sockets allow new IPv4 protocols to be implemented in user space.       A raw socket receives or sends the raw datagram not including link       level headers."
* https://tools.ietf.org/html/rfc3542#section-3 "When we say "an ICMPv6 raw socket" we mean a socket created by   calling the socket function with the three arguments AF_INET6,   SOCK_RAW, and IPPROTO_ICMPV6."

pfalcon commented:
Now regarding the implementation. I would consider one here "strange". Perhaps, "over-engineered". If we want to feed raw ethernet frames from eth driver(s) to packet socket, we just need to create a queue to store such packets for packet socket implementation to read them from. There's no need to do all those patchings in subsys/net/ip/. Because just as with CAN sockets, packet sockets have nothing to do with IP.

nashif commented:
given that this is being done for socketCAN, it is SOCK_RAW, see for example:

https://www.kernel.org/doc/Documentation/networking/can.txt
```
Like TCP/IP, you first need to open a socket for communicating over a
CAN network. Since SocketCAN implements a new protocol family, you
need to pass PF_CAN as the first argument to the socket(2) system
call. Currently, there are two CAN protocols to choose from, the raw
socket protocol and the broadcast manager (BCM). So to open a socket,
you would write::

    s = socket(PF_CAN, SOCK_RAW, CAN_RAW);

and::

    s = socket(PF_CAN, SOCK_DGRAM, CAN_BCM);
```

but will leave it to the original poster to answer the above.

pfalcon commented:
> given that this is being done for socketCAN

What's done for socketCAN? This is PR advertised as "network raw socket support", period. What's intended to build of it is irrelevant in the context of this PR. And such building-on, if any, can be erroneous. But that's also irrelevant so far, given the suspicion that this PR itself has enough mix-ups in it which need to be fixed first.

pfalcon commented:
Besides, the example you quote only confirms the conclusion I came to either: the top-level dispatch happens on the value of `family` - either AF_PACKET or AF_CAN (or PF_CAN, to mix up things a bit). Then, each socket family also dispatches on `socktype` - internally in its own implementation. Socket implementations of AF_PACKET vs AF_CAN would be separate, nothing is intended to depend on each other.

But indeed, let's hear counter-arguments of PR authors.

nashif commented:
> What's done for socketCAN? This is PR advertised as "network raw socket support", period.

this PR is to support the socket CAN work, hence the talk about RAW here makes sense, thought it could be on the can level and having CAN RAW socket, etc.

jukkar commented on include/net/net_pkt.h:
```diff
@@ -132,7 +132,7 @@ struct net_pkt {
 				 * Used only if defined(CONFIG_NET_ROUTE)
 				 */
-	u8_t family     : 4;	/* IPv4 vs IPv6 */
+	u8_t family     : 5;	/* IPv4 or IPv6 or AF_PACKET */
```
Indeed, I think we need to rewrite the actual values of the family in order to squeeze them into smallest amount of bits.

tbursztyka replied:
first patch of my PR #11775  is doing this, I can put it in its own PR to get applied asap if you want

pfalcon replied:
> I can put it in its own PR to get applied asap if you want

+1

jukkar replied:
> I can put it in its own PR to get applied asap if you want

Sounds like a good plan.

jukkar commented:
> this PR is to support the socket CAN work, hence the talk about RAW here makes sense, thought it could be on the can level and having CAN RAW socket, etc.

This PR is needed for socket CAN because:
* it provides the SOCK_RAW symbol that is used in socket CAN
* recv_raw() function in net_context.c can be shared by both

Other than that they (socket CAN and raw socket) are not really related to each other.

pfalcon commented:
> This PR is needed for socket CAN because:
> ...
> Other than that they (socket CAN and raw socket) are not really related to each other.

Thanks for clarification. My suggestions on specific points would be:

>    it provides the SOCK_RAW symbol that is used in socket CAN

Let's add that, along with SOCK_SEQPACKET, as a separate trivial PR. This way, we'll have definitions of all socket types mandated by POSIX, and can not return to it in the future.

>    recv_raw() function in net_context.c can be shared by both

Per my above sockets, I'd suggest to start development with not introducing dependencies among implementations of unrelated socket types. Instead, let's have good test coverage, so we can refactor implementations to improve code reuse later, once they're stable.

rveerama1 committed:
net: l2: ethernet: Handover net packets to driver

If packet family is AF_PACKET and CONFIG_NET_SOCKETS_PACKET
is enabled, just handover the packet to driver for sending.
L2 layer will not touch AF_PACKETs at the moment.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

dbkinder reviewed (commented):
No docs changes here to review.

rveerama1 committed:
net: core: Handle packets when packet sockets are enabled

If CONFIG_NET_SOCKETS_PACKET is enabled, then feed the packet
to net_packet_socket_input() for processing. It will search
for the net_contexts and if proper handler is found, pass
the packet to connection handler.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 commented:
> Now regarding the implementation. I would consider one here "strange". Perhaps, "over-engineered". If we want to feed raw ethernet frames from eth driver(s) to packet socket, we just need to create a queue to store such packets for packet socket implementation to read them from. There's no need to do all those patchings in subsys/net/ip/.

True, that's an ultimate goal.

rveerama1 commented:
Tried to address as many as I can. Please a take look again. Thanks.

rveerama1 commented on subsys/net/ip/connection.c:
```diff
@@ -798,14 +803,14 @@ static bool is_invalid_packet(struct net_pkt *pkt,
 	struct net_if *pkt_iface = net_pkt_iface(pkt);
 
-#if defined(CONFIG_NET_CONN_CACHE)
+#if defined(CONFIG_NET_CONN_CACHE) && defined(CONFIG_NET_UDP)
```
UDP or TCP. Have to investigate that support to RAW sockets.

rveerama1 commented:
recheck

pfalcon commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -119,6 +119,12 @@ int _impl_zsock_socket(int family, int type, int proto)
 #endif
 
+#if defined(CONFIG_NET_SOCKETS_RAW)
+	if (family == AF_PACKET && type == SOCK_RAW) {
```
Please dispatch only on AF_PACKET, handle type in your handler already. It would be pretty easy to support both SOCK_RAW and SOCK_DGRAM for AF_PACKET sockets, as described in http://man7.org/linux/man-pages/man7/packet.7.html

rveerama1 replied:
ok.

pfalcon commented on subsys/net/lib/sockets/sockets_internal.h:
```diff
@@ -49,4 +49,6 @@ struct socket_op_vtable {
 
 int ztls_socket(int family, int type, int proto);
 
+int zraw_socket(int family, int type, int proto);
```
Please rename to zpacket_socket()

pfalcon commented on subsys/net/lib/sockets/sockets_raw.c:
```diff
@@ -0,0 +1,351 @@
+/*
```
Please rename the file to sockets_packet.c

pfalcon reviewed (changes requested):
Thanks for the changes so far.

The biggest issue is however not fixed: based on the docs http://man7.org/linux/man-pages/man7/packet.7.html and previous discussion, this implements **packet sockets**, not *raw socketds*. Please update all the names and filenames accordingly. Thanks.
I'm helping in that regard by updating the PR title.

rveerama1 commented:
> The biggest issue is however not fixed: based on the docs http://man7.org/linux/man-pages/man7/packet.7.html and previous discussion, this implements **packet sockets**, not _raw socketds_. Please update all the names and filenames accordingly. Thanks.
> I'm helping in that regard by updating the PR title.

As per my understanding packet sockets are special type of Raw sockets (but still part of raw sockets). That's why I still kept the name as it is.

rveerama1 commented:
>   * The listen() and accept() functions are without any functionality,

As per my investigation in raw sockets there is no client and server roles. Based on that these calls are not needed.

pfalcon commented:
> As per my understanding packet sockets are special type of Raw sockets (but still part of raw sockets). That's why I still kept the name as it is.

It's vice-versa actually:

AF_INET sockets can be:
* SOCK_STREAM, SOCK_DGRAM, SOCK_SEQPACKET, SOCK_RAW

AF_INET6 sockets can be:
* SOCK_STREAM, SOCK_DGRAM, SOCK_SEQPACKET, SOCK_RAW

AF_PACKET sockets can be:
* SOCK_DGRAM, SOCK_RAW

AF_CAN sockets can be (https://www.kernel.org/doc/Documentation/networking/can.txt):
* SOCK_DGRAM, SOCK_SEQPACKET, SOCK_RAW

So, the main dispatch happens on the address family. Within sockets of the same address family, there're sockets of different types. Raw sockets of AF_INET are different matter comparing to raw sockets of AF_PACKET.

pfalcon commented:
>> The listen() and accept() functions are without any functionality,   

> As per my investigation in raw sockets there is no client and server roles. Based on that these calls are not needed.

So, there're 2 choices:

* Either remove those functions at all
* Or make them return error.

Returning 0 (successful return) is not ok.

rveerama1 commented:
> So, there're 2 choices:
> 
>     * Either remove those functions at all
> 
>     * Or make them return error.
> 
> 
> Returning 0 (successful return) is not ok.

ok, I will modify.

rveerama1 commented:
> It's vice-versa actually:
> 
> AF_INET sockets can be:
> 
>     * SOCK_STREAM, SOCK_DGRAM, SOCK_SEQPACKET, SOCK_RAW
> 
> 
> AF_INET6 sockets can be:
> 
>     * SOCK_STREAM, SOCK_DGRAM, SOCK_SEQPACKET, SOCK_RAW
> 
> 
> AF_PACKET sockets can be:
> 
>     * SOCK_DGRAM, SOCK_RAW
> 
> 
> AF_CAN sockets can be (https://www.kernel.org/doc/Documentation/networking/can.txt):
> 
>     * SOCK_DGRAM, SOCK_SEQPACKET, SOCK_RAW
> 
> 
> So, the main dispatch happens on the address family. Within sockets of the same address family, there're sockets of different types. Raw sockets of AF_INET are different matter comparing to raw sockets of AF_PACKET.

I am not saying that's not true.  What I am trying to explain is how it is implemented in zephyr or how am looking at.
We have
1) UDP / SOCK_DGRAM (ipv4, ipv6)
2) TCP / SOCK_STREAM (ipv4, ipv6)

In a similar way SOCK_RAW (AF_INET, AF_INET6, AF_PACKET) related functionality regardless of family will be in "raw" related filenames.

In other words Zephyr has support of
SOCK_DGRAM  (AF_INET, AF_INET6)
SOCK_STREAM (AF_INET, AF_INET6)

This version will add
SOCK_RAW (AF_PACKET).

Future support will be
SOCK_RAW (AF_INET or AF_INET6).

rveerama1 commented:
> Future support will be
> SOCK_RAW (AF_INET or AF_INET6).

And to support some of above, we might need to separate L2 or L3 or etc layers to reduce dependency. I have to investigate that.

pfalcon commented:
> What I am trying to explain is how it is implemented in zephyr or how am looking at.

Right. I, from my side, as a Socket Subsystem maintainer, trying to explain how it *should* be implemented.

> In a similar way SOCK_RAW (AF_INET, AF_INET6, AF_PACKET) related functionality regardless of family will be in "raw" related filenames.

That's not correct approach, per the details in above comments. Each address family should be implemented by a separate module (AF_INET and AF_INET6 being implemented by one module is the exception). These modules can be completely independent. They can be even dynamically loaded, once we have support dynaloaded modules. Each module processes different socket types (DGRAM vs RAW vs whatever) as the semantics of a particular address family requires (as an example again, RAW sockets for AF_INET are different from RAW sockets for AF_PACKET).

pfalcon commented:
> And to support some of above, we might need to separate L2 or L3 or etc layers to reduce dependency. I have to investigate that.

We absolutely need. AF_PACKET are strictly L2 sockets. AF_INET/AF_INET6 are strictly L3/L4 sockets. You shouldn't mix them up.

rveerama1 commented:
Solved rebase conflicts with master. Nothing else.

rveerama1 commented:
Working on rebase against master.

jukkar commented on subsys/net/ip/net_pkt.c:
```diff
@@ -550,10 +550,10 @@ static struct net_pkt *net_pkt_get(struct k_mem_slab *slab,
 #endif /* CONFIG_NET_CONTEXT_PRIORITY */
```
In commit "net: pkt: Set default data_len to iface MTU", the commit message is hard to understand. Perhaps something like this could be used:

`When creating net_pkt, the default value of net_pkt data_len was set to zero. In case of RAW sockets, when family is AF_PACKET, the data_len will be zero as there is no higher level protocol information. So in this case, the default value of data_len must be set to interface MTU.`

rveerama1 replied:
ok

jukkar commented on include/net/net_context.h:
```diff
@@ -878,6 +878,19 @@ int net_context_get_option(struct net_context *context,
+/**
+ * @brief Find matching network context with given input values.
+ *
+ * @param family IP address family (AF_INET or AF_INET6 or AF_PACKET)
```
As we are getting non IP protocols, better to say here `@param family Network protocol address family.`

rveerama1 replied:
ok.

jukkar commented on include/net/net_context.h:
```diff
@@ -878,6 +878,19 @@ int net_context_get_option(struct net_context *context,
+ *
+ * @param family IP address family (AF_INET or AF_INET6 or AF_PACKET)
+ * @param type Type of the socket, SOCK_STREAM or SOCK_DGRAM or SOCK_RAW
+ * @param proto protocol, IPPROTO_UDP, IPPROTO_TCP or etc.
```
Say here `@param proto Network protocol related to address family. Example: IPPROTO_UDP, IPPROTO_TCP etc.`

rveerama1 replied:
ok

jukkar commented on include/net/net_context.h:
```diff
@@ -878,6 +878,19 @@ int net_context_get_option(struct net_context *context,
+ * @brief Find matching network context with given input values.
+ *
+ * @param family IP address family (AF_INET or AF_INET6 or AF_PACKET)
+ * @param type Type of the socket, SOCK_STREAM or SOCK_DGRAM or SOCK_RAW
```
Too many "or" here, just say `@param type Type of the network socket. Values is either SOCK_STREAM, SOCK_DGRAM or SOCK_RAW`

rveerama1 replied:
ok

jukkar commented on subsys/net/ip/CMakeLists.txt:
```diff
@@ -27,6 +27,7 @@ zephyr_library_sources_ifdef(CONFIG_NET_STATISTICS   net_stats.c)
 zephyr_library_sources_ifdef(CONFIG_NET_TCP          connection.c tcp.c)
 zephyr_library_sources_ifdef(CONFIG_NET_TRICKLE      trickle.c)
 zephyr_library_sources_ifdef(CONFIG_NET_UDP          connection.c udp.c)
+zephyr_library_sources_ifdef(CONFIG_NET_SOCKETS_PACKET  connection.c)
```
This might give compiler error if TCP and SOCKETS_PACKET is defined at the same time (I did not test this).

jukkar commented on subsys/net/ip/net_context.c:
```diff
@@ -397,6 +425,17 @@ static int bind_default(struct net_context *context)
+
+		ll_addr.sll_family = AF_PACKET;
+		ll_addr.sll_protocol = ETH_P_ALL;
+		ll_addr.sll_ifindex = net_if_get_by_iface(net_if_get_default());
```
At some point in future, we might want to only use an interface that supports AF_PACKET. For that there would probably be a capability flag in net_if that would indicate the AF_PACKET support.

rveerama1 replied:
ok.

jukkar commented on subsys/net/ip/CMakeLists.txt:
```diff
@@ -27,7 +27,7 @@ zephyr_library_sources_ifdef(CONFIG_NET_STATISTICS   net_stats.c)
 zephyr_library_sources_ifdef(CONFIG_NET_TCP          connection.c tcp.c)
```
The commit message of "net: core: Handle packets when packet sockets are enabled " is difficult to read.
What about writing it like this:
`If CONFIG_NET_SOCKETS_PACKET is enabled, then feed the packet to net_packet_socket_input() for processing. It will search the net_contexts and if proper handler is found, pass the packet to connection handler.`

rveerama1 replied:
ok

jukkar commented on subsys/net/lib/sockets/CMakeLists.txt:
```diff
@@ -7,6 +7,7 @@ zephyr_sources(
   sockets_select.c
```
In commit "net: sockets: Add socket api to support AF_PACKET", there is difficult to understand sentence.
Could we say here:

```
This commit adds packet socket support to socket api. This version supports
basic packet socket features. Protocol family is AF_PACKET, type
of socket is SOCK_RAW and proto type is ETH_P_ALL. The user
will receive every packet (with L2 header) on the wire.
For TX, the subsystem expects that the user has set all the protocol headers (L2 and L3) properly.
```

jukkar commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -96,6 +96,18 @@ config NET_SOCKETS_OFFLOAD
+	  responsibility to provide all the headers (e.g L2, L3 and so on)
+	  while sending. While receiving, packets (including all the headers)
+	  will be feed to sockets as it as from the driver. This version
+	  supports only over Etherent.
```
Last two sentences are cryptic, could we say here:

```
When packets are received, the incoming data which includes all the protocol headers are feed to the listening socket.
This version supports only communication over Ethernet.
```

Because we only support Ethernet atm, please add  dependencies properly. So "depends on NET_L2_ETHERNET"

jukkar commented on samples/net/sockets/packet/README.rst:
```diff
@@ -0,0 +1,28 @@
+Building and Running
+********************
+
+When the application runs, it opens a packet socket and prints
```
This chapter sounds wrong. Perhaps something like:

```
When the application is run, it opens a packet socket and prints 
the length of the packet it receives. After that it sends a dummy packet
every 5 seconds. You can use Wireshark to observe these sent and received packets.
```

jukkar commented on samples/net/sockets/packet/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+  test:
+    harness: net
+    platform_whitelist: native_posix
+    tags: net sockets
```
Add also "packet-socket" tag

rveerama1 commented:
Rebased and addressed comments.

pfalcon commented on include/net/net_ip.h:
```diff
@@ -128,14 +131,65 @@ struct sockaddr_in_ptr {
+#define PACKET_FASTROUTE    6
+
+#if defined(CONFIG_NET_IPV4)
+#undef NET_SOCKADDR_MAX_SIZE
```
I'd say it's a bit strange to have these #undef's. In such way, we can redefine these values in loop, and won't have a hint where the issues happens. Can't we just define these values consistently in one chunk of preprocessor code, without #ifdef's. @jukkar, @tbursztyka: Please have a look, I'd be concerned in your place.

jukkar replied:
The undef's are there in order to avoid preprocessor error if multiple protocols are defined. The idea is that the largest size sockaddr is the last one.

pfalcon replied:
> The idea is that the largest size sockaddr is the last one.

I'm not sure how obvious is that idea. It's definitely not documented in a code comment before that chunk, so someone adding a new sockaddr struct can easily stuff it in a random place in sequence.

rveerama1 replied:
Adding comments are enough? Is there anything to do further?

pfalcon replied:
I'd say yes, if it's documented in a comment before that block, it's fine (I assume @jukkar introduced it in his initial version).

rveerama1 replied:
> I assume @jukkar introduced it in his initial version

yes.

rveerama1 replied:
> I'd say yes, if it's documented in a comment before that block, it's fine

Done.

pfalcon commented on subsys/net/ip/connection.c:
```diff
@@ -777,6 +782,12 @@ static bool is_invalid_packet(struct net_pkt *pkt,
 		}
 	}
 
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_PACKET)) {
```
Can you please add a code comment what happens here? (I suspect the root problem is that the function is misnamed, it should be "is_invalid_incoming_packet()" instead, but we have what we have, then comments are really helpful/needed.)

pfalcon commented on subsys/net/ip/connection.c:
```diff
@@ -777,6 +782,12 @@ static bool is_invalid_packet(struct net_pkt *pkt,
 	}
 
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_PACKET)) {
+		if (net_pkt_family(pkt) == AF_PACKET) {
```
Can you please add a code comment what happens here? (I suspect the root problem is that the function is misnamed, it should be "is_invalid_incoming_packet()" instead, but we have what we have, then comments are really helpful/needed.)

rveerama1 replied:
It's rebase issue. After rebasing with master (after Tomasz bit patchset) I had to introduce this. But it should be in different patch (I wrongly squashed it). Anyway I will add comments there.

pfalcon commented on subsys/net/ip/net_context.c:
```diff
@@ -399,6 +427,17 @@ static int bind_default(struct net_context *context)
+
+		ll_addr.sll_family = AF_PACKET;
+		ll_addr.sll_protocol = ETH_P_ALL;
+		ll_addr.sll_ifindex = net_if_get_by_iface(net_if_get_default());
```
Do you remember, that in POSIX, interface index can't be 0 (zero means "doesn't care"). That effectively means that indexes start from 1. So, this is not correct.

rveerama1 replied:
> Do you remember, that in POSIX, interface index can't be 0 (zero means "doesn't care").

Don't know about that.

rveerama1 replied:
> That effectively means that indexes start from 1. So, this is not correct.

i.e. we have to change our indexing from 1.

pfalcon replied:
> Don't know about that.

Well, please read up on that in http://man7.org/linux/man-pages/man7/packet.7.html :

> sll_ifindex is the interface index of the interface (see
          netdevice(7)); 0 matches any interface (only permitted for bind‐
          ing).  sll_hatype is an ARP type as defined in the
          <linux/if_arp.h> include file.


---
> i.e. we have to change our indexing from 1.

Yes, we would need to +1 internal Zephyr number when we pass it to *socket* functions, and -1 what we get from user via *socket* functions. I'm just not sure at what level to do that, but given that we *have* to do that, butter just do that in net_context functions, to avoid having different indexes in different API. @jukkar, @tbursztyka,  maybe you'd have suggestion here.

jukkar replied:
I noticed the same thing that linux CAN manual pages talk about index 0 being special. Not sure how we could accomplish that currently, it sounds very error prone if we start to +-1 the interface index value in net_context. Probably the best option is to make net_if_get_by_index() do the -1 and net_if_get_by_iface() to return +1 for the network interface index value and make sure that no code accesses network interface array directly. I would leave this issue after 1.14 merge window closes.

pfalcon replied:
> Probably the best option is to make net_if_get_by_index() do the -1 and net_if_get_by_iface() to return +1 for the network interface index value and make sure that no code accesses network interface array directly.

I can only +1 such a solution: it's arguably a wart, I don't like 1-based indexing myself, but if POSIX specifies that (and likely it's not in POSIX explicitly, it's just how BSD did it, and all UNIXes adhere to), it's better to adopt that scheme as native to Zephyr too.

While it's not pretty formal decision, pragmatically I agree that we probably could live with this "numbering bug" for some time (i.e. not do anything in this patch), if we don't forget to fix it afterwards.

pfalcon replied:
> Is there anything to do further? shall I mark this conversation resolved?

I ticketed that as https://github.com/zephyrproject-rtos/zephyr/issues/13084, and in the circumstances we have, I agree it's "ok" to keep stuff as is here.

rveerama1 replied:
Ok, thank you.

pfalcon commented on subsys/net/ip/net_context.c:
```diff
@@ -1087,6 +1182,7 @@ static int sendto(struct net_pkt *pkt,
 
 #if defined(CONFIG_NET_IPV4)
 	if (net_pkt_family(pkt) == AF_INET) {
+
```
Spurious changes, please remove.

rveerama1 replied:
ok

pfalcon commented on subsys/net/ip/net_context.c:
```diff
@@ -1098,6 +1194,7 @@ static int sendto(struct net_pkt *pkt,
 		}
 	} else
 #endif /* CONFIG_NET_IPV4 */
+
```
Spurious changes, please remove.

rveerama1 replied:
ok

pfalcon commented on subsys/net/ip/net_context.c:
```diff
@@ -1175,6 +1272,12 @@ int net_context_send(struct net_pkt *pkt,
 
 	k_mutex_lock(&context->lock, K_FOREVER);
 
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_PACKET) &&
```
Please add a comment that only sendto() is supported for AF_PACKET sockets.

rveerama1 replied:
ok

pfalcon commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -96,6 +96,19 @@ config NET_SOCKETS_OFFLOAD
+	  responsibility to provide all the headers (e.g L2, L3 and so on)
+	  while sending. While receiving, packets (including all the headers)
+	  will be feed to sockets as it as from the driver. This version
+	  supports only over Etherent.
```
> This version supports only over Etherent.

This sentence also needs a grammar and typo fix up. E.g. "This version supports only Ethernet L2", or similar.

rveerama1 replied:
fixing type, I don't see issue in rest.

pfalcon replied:
"supports only over" is not proper literature English. It should be "supports &lt;something> only over", or should be rephrased completely.

rveerama1 replied:
Added "depends on NET_L2_ETHERNET" to config option. Hope that's fine.

pfalcon commented:
@rveerama1: Thanks for all the refactoring so far, this shapes up really well and now looks *much* cleaner! Still some issues to address (sorry for posting as individual comments, will try to do that in batched reviews next time).

rveerama1 commented:
@dbkinder can you please go through again?

pfalcon commented:
Ok, this subthread is no longer visible by default, so let me bring it to the top level, as it's pretty important:

(The talk is about net_context_find() function which was previously in this PR)
https://github.com/zephyrproject-rtos/zephyr/pull/12564#discussion_r253480160

@rveerama1:

> Hmm, after all this API is not required at the moment. It was introduced for a reason. If user open a packet socket and udp/tcp socket. The same net_pkt has to be feed to both the sockets. 

Right, so I wondered if this PR actually does that.

> So if user enabled CONFIG_NET_SOCKETS_PACKET and opened a packet socket, then packet has to be feed as it it. Then for udp socket data has to feed. May be we have to clone the packet.

We definitely need to do something about this case. But cloning a packet should be the last, "nothing else worked" solution. Because we have a packet reference  count exactly for such cases! If we don't make use of it, why do we need it at all? So, the first choice would be to just increment refcount for each queue the packet gets added to (packet socket vs normal socket).

Problem: in many places in the stack a code deals in such a way as if it owns a packet completely - without actually checking refcount. @tbursztyka currently fixes a case with ethernet header, in https://github.com/zephyrproject-rtos/zephyr/pull/12709#issuecomment-460751521 I admit that socket TCP code is written with similar assumptions. We to go thru the stack and fix such cases.

> But If user enabled CONFIG_NET_SOCKETS_PACKET but doesn't opened any packet socket, then we need not to clone. So to find that particular context we need this helper.

> After all in this initial version we don't need this at all. I will remove that patch.

Thanks, the smaller the initial code, the better.

rveerama1 commented:
> We definitely need to do something about this case. But cloning a packet should be the last, "nothing else worked" solution. Because we have a packet reference count exactly for such cases! If we don't make use of it, why do we need it at all? So, the first choice would be to just increment refcount for each queue the packet gets added to (packet socket vs normal socket).

Packet socket expects packet with L2 header, normal socket expects only data (we have to remove L2 and L3 header). How could refcount mechanism works here with same net_pkt?

pfalcon commented:
> Packet socket expects packet with L2 header, normal socket expects only data (we have to remove L2 and L3 header). How could refcount mechanism works here with same net_pkt?

That's more a question to @tbursztyka. What you write a requirement, he can tell us how to implement it. An obvious way is to provide "current position" pointer. Normal sockets move that pointer past the headers, packet sockets have it at the headers. The main point that moving pointer should be non-destructive in case of refcount > 1, and can be destructive if recount == 1 (these rules are ok for recv packets only of course, and with some additional rules of how refcounts are initially set up).

jukkar reviewed (approved):
I think this is good enough for merging, we can fix any issues later.

tbursztyka commented on subsys/net/lib/sockets/sockets_packet.c:
```diff
@@ -0,0 +1,339 @@
+		recv_len = max_len;
+	}
+
+	net_frag_linearize(buf, recv_len, pkt, 0, recv_len);
```
Use the new API directly, net_pkt_read_new(). Check what I have done in #12989 (see the 2 last commits on udp, tcp for socket). 
(The only trick is to have a backup of the cursor in case ZSOCK_MSG_PEEK)

rveerama1 replied:
According to your suggestion, if packet has 200 bytes, and caller tries to read (recvfrom) 100 bytes, means **error**. I am bit confused.

tbursztyka replied:
why? you did already the right check above line 216. You have to keep that of course.

rveerama1 replied:
Ok.

rveerama1 replied:
Done, please take a look again.

tbursztyka reviewed (commented):
Just one tiny change in socket side. rest is good to me.

pfalcon reviewed (approved):
Well, I agree that @rveerama1 did a great job on refactoring this and addressing comments. My initial intention was to make sure there's no obviously wrong stuff, and just be +0, on it, because I still think that we're hasting with pushing this stuff into 1.14. But well, again, in circumstances we have, it's really good job, so let me be +1.

Major issue remains the sample - right from the start it was requested to follow a simple rule "If it's socket API, please provide a sample which works both on Linux and Zephyr". And yet the sample remains Zephyr-specific. @rveerama1, please rewrite it completely as the next step, remove threads, etc. No, you can use pthreads of course, but I don't see why they're needed for just a sample. How true example of POSIX/BSD portability is truly a must.

rveerama1 commented:
> Major issue remains the sample - right from the start it was requested to follow a simple rule "If it's socket API, please provide a sample which works both on Linux and Zephyr". And yet the sample remains Zephyr-specific. @rveerama1, please rewrite it completely as the next step, remove threads, etc. No, you can use pthreads of course, but I don't see why they're needed for just a sample. How true example of POSIX/BSD portability is truly a must.

Ok, I will modify in another patch. Thanks.

tbursztyka reviewed (changes requested):
missing cursor backup in case of flasg & ZSOCK_MSG_PEEK

rveerama1 commented on subsys/net/lib/sockets/sockets_packet.c:
```diff
@@ -0,0 +1,342 @@
+	size_t recv_len = 0;
+	s32_t timeout = K_FOREVER;
+	struct net_pkt *pkt;
+
```
Ok, done and sorry. Looks like it takes some to get familiar with new api.

tbursztyka replied:
yep, but hopefully it's much simpler than legacy one. have a look at tests/net/net_pkt_new/, I hope it answers most of your question there.
